Domain Picker: refactor folder structure

ciampo opened the pull request:
#### Changes proposed in this Pull Request

* Refactor the structure of the Domain Picker `src` folder to allow for better modularity and tests granularity
* Refer to #51144 and to this PR's commit log for details list of changes

#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

These changes should only affect the internal structure of the Domain Picker package, and therefore shouldn't affect any user-facing behaviour

* Run `yarn && yarn start` and make sure that the project builds correctly
* Run `cd apps/editing-toolkit && yarn build` and make sure that the project builds correctly
* Run `cd packages/domain-picker && yarn test` and make sure that all tests pass (the `<button> cannot appear as a descendant of <button>.` error is expected)
* Smoke test Gutenboarding and Focused Launch

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Closes #51144

ciampo committed:
Move all domain picker components into src/components folder:

- Added the domain-suggestion-item folder, with an index.ts file
- Split domain suggestion item types to a separate file

ciampo committed:
Temporarily disable wpcalypso/jsx-classname-namespace rule

ciampo committed:
Move utils.ts to utils/index.ts

ciampo committed:
Update jest config to match tests from __test__ folder for domain picker package

ciampo committed:
Move isGoodDefaultDomainQuery unit tests to utils test folder

ciampo committed:
Move use your domain item and suggestion item tests to the component's __test__ folder

ciampo committed:
Cleanup duplicate test file

ciampo committed:
Move integration test to src/__test__ folder

ciampo committed:
Move the matchMedia mock to the __mock__ folder

ciampo committed:
Fix types patch import

ciampo committed:
Fix jest match patter from __tests__ to __test__

ciampo committed:
Add placeholder test files

ciampo committed:
Move mock domain suggestion data to __mock__ folder

ciampo commented on packages/domain-picker/jest.config.js:
```diff
@@ -1,5 +1,6 @@
 module.exports = {
 	preset: '../../test/packages/jest-preset.js',
 	testEnvironment: 'jsdom',
+	testMatch: [ '<rootDir>/**/__test__/**/*.[jt]s?(x)', '!**/.eslintrc.*' ],
```
Override default Calypso's config, so that we can put our tests in folders named `__test__`

StefanNieuwenhuis replied:
Maybe a bit of [nitpicking/] since Calypso is overruling the default Jest config anyway, but Jest's default is `__tests__` (plural) (see: https://jestjs.io/docs/configuration#testmatch-arraystring). The same goes for `__mock__` vs. `__mocks__`.

WDYT? Do we need to use plural too?

ciampo replied:
Updated the original issue's A/C and done across these commits:
- https://github.com/Automattic/wp-calypso/pull/51235/commits/cd951f36e0d91f5c4daea4f8571c9c81caf950f5
- https://github.com/Automattic/wp-calypso/pull/51235/commits/a5c22b64f101483b8e32811a250a4865bed11d47
- https://github.com/Automattic/wp-calypso/pull/51235/commits/eea4c35d2ebe964020cbecdd29e5759b8566563a
- https://github.com/Automattic/wp-calypso/pull/51235/commits/512ee17ce81e4ea1c82ab32dd4b95fd2d59e108d

ciampo commented on packages/domain-picker/src/components/domain-suggestion-item/__test__/suggestion-item.tsx:
```diff
@@ -12,8 +8,8 @@ import { render, fireEvent, screen } from '@testing-library/react';
-import '../matchMedia.mock';
-import SuggestionItem from '../src/domain-picker/suggestion-item';
+import '../../../__mock__/matchMedia.mock';
+import SuggestionItem from '../suggestion-item';
```
@StefanNieuwenhuis FYI — this will likely cause merge conflicts with #51177

StefanNieuwenhuis replied:
It certainly will and we have to use the DomainSuggestions mock there as well. So we have to discuss our merging strategy. 

I opt for merging this one first and rebase, update & merge #51177 after.

ciampo committed:
Add missing wpcalypso/jsx-classname-namespace disable rule

ciampo committed:
Move SUGGESTION_ITEM_TYPE back to suggestion-item.tsx

- Rename variables for consistency
- Always use variables (instead of raw value) across the codebase

ciampo commented on packages/domain-picker/src/components/domain-suggestion-item/__test__/suggestion-item.tsx:
```diff
@@ -1,7 +1,3 @@
-/**
- * @jest-environment jsdom
- */
-
```
Not necessary since it's already specified in this package's jest config

ciampo commented on packages/domain-picker/src/hooks/__test__/use-domain-suggestions.ts:
```diff
@@ -0,0 +1,5 @@
+describe( 'useDomainSuggestions hook', () => {
+	it( 'should be replaced with real tests', () => {
+		expect( 1 ).toBeTruthy();
+	} );
+} );
```
Opened #51242 to track Domain Picker hooks unit tests

ciampo commented on packages/domain-picker/src/hooks/__test__/use-persistent-selected-domain.ts:
```diff
@@ -0,0 +1,5 @@
+describe( 'usePersistentSelectedDomain hook', () => {
+	it( 'should be replaced with real tests', () => {
+		expect( 1 ).toBeTruthy();
+	} );
+} );
```
Opened #51242 to track Domain Picker hooks unit tests

ciampo commented on packages/domain-picker/src/utils/__test__/is-good-default-domain-query.ts:
```diff
@@ -1,7 +1,7 @@
  * Internal dependencies
  */
-import { isGoodDefaultDomainQuery } from '../src/utils';
+import { isGoodDefaultDomainQuery } from '../';
```
Missing utils unit tests tracked in #51243

ciampo commented on packages/domain-picker/src/components/__test__/domain-picker.tsx:
```diff
@@ -0,0 +1,5 @@
+describe( '<DomainPicker />', () => {
+	it( 'should be replaced with real tests', () => {
+		expect( 1 ).toBeTruthy();
+	} );
+} );
```
Missing domain picker components unit tests tracked in #51244

ciampo committed:
Split type and const exports (may fix build error)

StefanNieuwenhuis commented on packages/domain-picker/src/__mock__/suggestions.ts:
```diff
@@ -0,0 +1,12 @@
+ */
+import type { DomainSuggestion } from '@automattic/data-stores/src/domain-suggestions';
+
+export const MOCK_DOMAIN_SUGGESTION: DomainSuggestion = {
```
I think we have to extend and use this mock in #51177 too.

StefanNieuwenhuis commented on packages/domain-picker/src/components/domain-suggestion-item/suggestion-item-wrapper.tsx:
```diff
@@ -6,7 +6,7 @@ import * as React from 'react';
  * Internal dependencies
  */
-import type { SUGGESTION_ITEM_TYPE } from 'src';
+import { SUGGESTION_ITEM_TYPE, SUGGESTION_ITEM_TYPE_BUTTON } from './suggestion-item';
```
What do you think about this, more explicit, though verbose, approach?

```suggestion
import type { SUGGESTION_ITEM_TYPE } from './suggestion-item';
import { SUGGESTION_ITEM_TYPE_BUTTON } from './suggestion-item';
```

ciampo replied:
Done in https://github.com/Automattic/wp-calypso/pull/51235/commits/693ce38a124aa02293aeb15367692937c823995f

StefanNieuwenhuis commented on packages/domain-picker/src/components/domain-suggestion-item/suggestion-item-wrapper.tsx:
```diff
@@ -20,7 +20,7 @@ type WrappingComponentProps = WrappingComponentAdditionalProps &
 const WrappingComponent = React.forwardRef< HTMLButtonElement, WrappingComponentProps >(
 	( { type, disabled, ...props }, ref ) => {
-		if ( type === 'button' ) {
+		if ( type === SUGGESTION_ITEM_TYPE_BUTTON ) {
```
[nitpick/] This is scope creep since it doesn't have anything to do with folder restructuring.

ciampo replied:
That's partially true, although the reason these changes are in this PR is because initially I had extracted these variables to a separate file during the folder restructure, and then I changed my mind and moved them back.

It's a small scope creek but probably we can keep it in, since the changes are not affecting the functionality of the component?

StefanNieuwenhuis replied:
Thanks for elaborating. Although we have to consider it as scope creep, we can keep it in since the changes are not affecting the functionality of the component

StefanNieuwenhuis commented on packages/domain-picker/src/components/domain-suggestion-item/suggestion-item.tsx:
```diff
@@ -23,15 +25,16 @@ import { sprintf } from '@wordpress/i18n';
-export const ITEM_TYPE_BUTTON = 'button';
-export const ITEM_TYPE_INDIVIDUAL_ITEM = 'individual-item';
+export const SUGGESTION_ITEM_TYPE_RADIO = 'radio';
+export const SUGGESTION_ITEM_TYPE_BUTTON = 'button';
+export const SUGGESTION_ITEM_TYPE_INDIVIDUAL = 'individual-item';
 export type SUGGESTION_ITEM_TYPE =
-	| typeof ITEM_TYPE_RADIO
-	| typeof ITEM_TYPE_BUTTON
-	| typeof ITEM_TYPE_INDIVIDUAL_ITEM;
+	| typeof SUGGESTION_ITEM_TYPE_RADIO
+	| typeof SUGGESTION_ITEM_TYPE_BUTTON
+	| typeof SUGGESTION_ITEM_TYPE_INDIVIDUAL;
+
```
[nitpick/] This is scope creep since it doesn't have anything to do with folder restructuring.

ciampo replied:
See https://github.com/Automattic/wp-calypso/pull/51235#discussion_r599720224

StefanNieuwenhuis commented on packages/domain-picker/src/components/domain-suggestion-item/suggestion-item.tsx:
```diff
@@ -23,15 +25,16 @@ import { sprintf } from '@wordpress/i18n';
 import InfoTooltip from '../info-tooltip';
 import WrappingComponent from './suggestion-item-wrapper';
 // TODO: remove when all needed core types are available
```
Do we need an issue for this?

ciampo replied:
Opened #51328 to track it

StefanNieuwenhuis reviewed (commented):
Thanks for this PR @ciampo! It looks so structured already and I've tested it successfully! 

My only concern is the placeholder unit tests. Since we're not sure if we need to create unit tests for all (sub)components or that we can cover them with integration tests, I think that they should be left out. WDYT?

I left some inline comments as well. Can you please address them?

ciampo committed:
Split type imports from other imports

ciampo committed:
Remove placeholder test files / directories

ciampo committed:
Disable jest/no-mocks-import eslint rule for this package

ciampo committed:
Update configurations  to use __tests__ folder instead of __test__

ciampo committed:
Move __mock__/* and __test__/* to __mocks__/* and __tests__/*

ciampo committed:
Import useDomainSuggestions only as a type in tests

ciampo commented:
Hey @StefanNieuwenhuis, I addressed all of your comments. Can you have another look?


> My only concern is the placeholder unit tests. Since we're not sure if we need to create unit tests for all (sub)components or that we can cover them with integration tests, I think that they should be left out. WDYT?

Removed placeholder tests in https://github.com/Automattic/wp-calypso/pull/51235/commits/99a2df3e926af3b361c092537f53987df20f7812

StefanNieuwenhuis reviewed (approved):
Code & Tests LGTM! 🚢
